modules/flowable-engine/src/main/java/org/flowable/engine/impl/bpmn/helper/ClassDelegate.java

```diff
@@ -216,7 +217,15 @@ public void trigger(DelegateExecution execution, String signalName, Object signa
         }
 
         if (activityBehaviorInstance instanceof TriggerableActivityBehavior) {
-            ((TriggerableActivityBehavior) activityBehaviorInstance).trigger(execution, signalName, signalData);
+            try {
+                ((TriggerableActivityBehavior) activityBehaviorInstance).trigger(execution, signalName, signalData);
+            } catch (BpmnError error) {
+                ErrorPropagation.propagateError(error, execution);
+            } catch (RuntimeException e) {
+                if (!ErrorPropagation.mapException(e, (ExecutionEntity) execution, mapExceptions)) {
+                    throw e;
+                }
+            }
             if (triggerable) {
                 leave(execution);
             }
```

Comment:
This here is a bit strange how it works. If an BPMN error is thrown or the exception is somehow mapped then no exception is thrown and the leave will be reached. That could lead to some strange state I think. Can you please add test cases that cover this as well and make sure that we don't leave when an error happens.

Reply:
Hi @filiphr.

Currently `TakeOutgoing...Operation` is planned, but it is not executed because of
```
            // Execute the operation if the operation has no execution (i.e. it's an operation not working on a process instance)
            // or the operation has an execution and it is not ended
            if (operation.getExecution() == null || !operation.getExecution().isEnded()) {
```
In `CommanInvoker`.

I've put `leave(execution)` into try block to do not even plan to leave.

Reply:
That is really tangled @martin-grofcik. The fact that it works might be by chance. We shouldn't plan a leave if we don't need to leave through the flow. I'll have a look at your changes

Reply:
I checked your last changes and you still have not moved this `leave`.